src/main/java/org/cryptomator/frontend/fuse/locks/PathLock.java

```diff
@@ -1,12 +1,66 @@
 package org.cryptomator.frontend.fuse.locks;
 
-public interface PathLock extends AutoCloseable {
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.Unmodifiable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-	DataLock lockDataForReading();
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.function.Function;
 
-	DataLock lockDataForWriting();
+public record PathLock(@Unmodifiable List<String> pathComponents, @Nullable PathLock parent, ReadWriteLock rwLock, Lock lock, Function<List<String>, ReadWriteLock> dataLockSupplier) implements AutoCloseable {
```

Comment:
if i am seeing this correctly, the rwLock ist not actually used/needed in the record, right?

It is only used during the construction, which happen in the static factory methods.

Reply:
This requires explanation:

Since we're storing the ReadWriteLock in the LockManager's cache as a weak reference, we need to make sure to keep a strong reference as long as the high-level lock object is in use.

It _might_ be possible that the read and write locks returned by the ReadWriteLock have such a reference, but we better not rely on such an implementation detail.

I will add a inline-comment as well.

Reply:
fixed in c1c5d36